talk/expert/sfinae.tex

```diff
@@ -236,22 +236,29 @@
 \end{frame}
 
 \begin{frame}[fragile]
-  \begin{exampleblock}{Gaudi usage example}
+  \begin{exampleblockGB}{Usage example}{https://godbolt.org/z/Wc35ec8bx}{\texttt{enable\_if}}
+    \small
     \begin{cppcode*}{}
-      constexpr struct deref_t {
-        template
-          <typename In,
-           typename = typename std::enable_if_t
-                      <!std::is_pointer_v<In>>>
-        In& operator()(In& in) const { return in; }
+    template <typename T> constexpr bool is_iterable ... ;
 
-        template <typename In>
-        In& operator()(In* in) const {
-          assert(in!=nullptr); return *in;
+    struct Printer {
+        template<typename T,
+                 typename std::enable_if_t
+                   <!is_iterable<T>, bool> = true>
+        void print(T const& t) {
+            std::cout << t << "\n";
+        }
+        template<typename T,
+                 typename std::enable_if_t
+                   <is_iterable<T>, bool> = true>
+        void print(T const& t) {
+            for (auto const& item : t) {
+                std::cout << item << "\n";
+            }
         }
-      } deref;
+    };
```

Comment:
I would remove the outer struct `Printer`. And to implement such a function `print` I would use `if constexpr` here :)

Reply:
Indeed. So I've remove the struct.

For the if constexpr, yes, indeed and that's the point of a further slide. I've added a note at the bottom though and implemented both solutions in the godbolt example.